src/keyboardNavigation.ts

```diff
@@ -0,0 +1,82 @@
+import { isDeepestVisible, isAncestor, isEmptyObject } from './utils';
+
+function findDeepestVisibleChild(node: HTMLElement): HTMLElement {
+    if (isDeepestVisible(node)) return node;
+    const deeperObject = <HTMLElement>node.querySelector('[part=object]');
+    const nextDeepest = getLastChild(deeperObject);
+    if (!nextDeepest) return node;
+    return findDeepestVisibleChild(nextDeepest);
+}
+
+function findNearestParentWithAdditionalLeaves(node: HTMLElement, initial = node): HTMLElement {
+    const ul = node.parentElement;
+    if (!ul) return node;
+    const parent = ul.parentElement;
+    if (!parent || !parent.parentElement) return initial;
+    const parentAndSiblings = [...parent.parentElement.children];
+    if (isLastChild(parent, parentAndSiblings)) {
+        return findNearestParentWithAdditionalLeaves(parent, initial);
+    } else {
+        return getNextSibling(parent, parentAndSiblings);
+    }
+}
+
+export const navigateUp = (path: string, shadowRoot: ShadowRoot): void => {
+    const currentFocus = <HTMLElement>shadowRoot.querySelector(`[data-path="${path}"]`);
+    if (!currentFocus || !currentFocus.parentNode) return;
+    const currentFocusAndSiblings = [...currentFocus.parentNode.children];
+    const previousSibling = getPreviousSibling(currentFocus, currentFocusAndSiblings);
+    if (previousSibling instanceof ShadowRoot) return;
+    const newFocus =
+        !isDeepestVisible(previousSibling) && !isAncestor(path, previousSibling.getAttribute('data-path') || '')
+            ? findDeepestVisibleChild(previousSibling)
+            : previousSibling;
+    if (newFocus) changeFocusedKey(newFocus, currentFocus);
+};
+
+export const navigateDown = (path: string, shadowRoot: ShadowRoot): void => {
+    const currentFocus = <HTMLElement>shadowRoot.querySelector(`[data-path="${path}"]`);
+    if (!currentFocus || !currentFocus.parentNode) return;
+    const currentFocusAndSiblings = [...currentFocus.parentNode.children];
+    const newFocus = findNextElementToFocus(currentFocus, currentFocusAndSiblings);
+    if (newFocus) changeFocusedKey(newFocus, currentFocus);
+};
+
+const findNextElementToFocus = (currentFocus: HTMLElement, currentFocusAndSiblings: Array<Element>) => {
+    if (!isDeepestVisible(currentFocus)) {
+        if (isEmptyObject(currentFocus)) {
```

Comment:
i'd have thought an empty object was always "the deepest visible", but that would make this block unreachable. what's deeper?

Reply:
You would be correct to assume that, I did too. Unfortunately the rendering of these nodes does not account for empty object being different than non-empty, so I'm handling in a custom way, and I think that makes sense to express independently.

No changes for now, but there is more to be discussed about `isDeepestVisible`. I'm not writing off adding `|| isEmptyObject(node)` to `isDeepestVisible` yet.

Reply:
to me it feels like an empty object is the deepest visible one, but i might not have a clear image there